src/metabase/db/connection.clj

```diff
@@ -143,7 +143,10 @@
                     (.commit connection))
                   result)
                 (catch Throwable e
-                  (.rollback connection savepoint)
+                  (try
+                    (.rollback connection savepoint)
+                    (catch Exception rollback-e (throw (ex-info (ex-message e) {:cause e :rollback-error rollback-e}))))
```

Comment:
This line is getting pretty long. I'd break it up a bit, and also suggest this construction:
rename `e` to `txn-e` and then do
```clojure
(catch Throwable txn-e
  (try (.rollback connection savepoint)
       (catch Exception rollback-e
         (throw (ex-info "Error in transaction"
                         {:rollback-error rollback-e}
                         txn-e))))
  (throw txn-e))
```

And in the tests:

```clojure
(deftest rollback-error-handling
  (testing "rollback error handling"
    (let [real-conn (.getConnection ^DataSource mdb.connection/*application-db*)
          mock-conn (reify Connection
                      (rollback [_ _savepoint]
                        (throw (ex-info "Rollback error" {})))
                      (setAutoCommit [_ auto?]
                        (.setAutoCommit real-conn auto?))
                      (getAutoCommit [_]
                        (.getAutoCommit real-conn))
                      (setSavepoint [_]
                        (.setSavepoint real-conn))
                      (commit [_]
                        (.commit real-conn)))]
      (binding [t2.connection/*current-connectable* mock-conn]
        (let [e (is (thrown? Exception
                             (t2/with-transaction [_t-conn] (throw (ex-info "Actual error" {})))))]
          (def e e)
          (is (= "Error in transaction" (ex-message e)))
          (is (= "Rollback error" (-> e ex-data :rollback-error ex-message)))
          (is (= "Actual error" (-> e ex-cause ex-message))))))))
```

Reply:
Mostly used your suggestion, what do you think of the new version?